lib/dialects/abstract/query-generator.js

```diff
@@ -1387,11 +1387,20 @@ module.exports = (function() {
       }
 
       if (options.lock && this._dialect.supports.lock) {
-        if (options.lock === 'SHARE') {
+        var lock = options.lock;
+        if (typeof options.lock === 'object') {
+          lock = options.lock.level;
+        }
+        if (this._dialect.supports.lockKey && (lock === 'KEY SHARE' || lock === 'NO KEY UPDATE')) {
+          query += ' FOR ' + lock;
+        } else if (lock === 'SHARE') {
           query += ' ' + this._dialect.supports.forShare;
         } else {
           query += ' FOR UPDATE';
         }
+        if (this._dialect.supports.lockOf && options.lock.of instanceof Model) {
+          query += ' OF ' + this.quoteTable(options.lock.of.name);
```

Comment:
@BridgeAR explained on IRC that LOCK OF operates on the aliased table, and we alias to `model.name` so whats in the PR is correct.